### ISSUE TITLE
[critical] `pgtrigger.ignore` persists outside of its context if still in same transaction

### ISSUE THREAD

**Issue by ralokt** [Opus10/django-pgtrigger#49]
# tl;dr
`pgtrigger` uses `SET LOCAL` to set the SQL variables controlling trigger suppression. Changes made by `SET LOCAL` are transaction-scoped, meaning that they leak into following queries in the same transaction. Since exiting the last ignore context manager also removes the `pre_execute_hook`, the new/empty value for `pgtrigger.ignore` isn't injected into any queries, leaving it where it was inside the `ignore` context block.

**This can easily lead to triggers being ignored entirely**. If there is an official interface enforced by write protection, that write protection is gone after you use it until the transaction is over. If you wrap every request in a transaction, the write protection is gone for the rest of the request.

I will edit this issue to provide a more detailed explanation, but want to publish it ASAP as I think it is critical.

EDIT:
Firstly, I hope it's OK that I marked this issue as critical. Given that this library is meant to provide reliable hard guarantees, and given that this bug makes it so that said hard guarantees are quietly bypassed, I felt that this was appropriate.

# Context

I'll briefly recap how `pgtrigger` (and more specifically the mechanism to temporarily ignore triggers) works so that I can explain this issue and discuss potential solutions in context.

The triggers created by `pgtrigger` are functions that, before performing the action they usually would, check if they should run or not in the first place. They do this by reading `pgtrigger.ignore`, an array of triggers that should currently be ignored, and check whether they are in that array. If so, they perform no additional action.

So, by setting `pgtrigger.ignore`, we can temporarily disable triggers, giving us a low-level interface.

The high-level interface is `pgtrigger.core.ignore`, a context manager that temporarily sets `pgtrigger.ignore` to its argument list for queries that happen inside its context block.

It achieves this by
 - remembering the currently ignored triggers in a thread-local variable
 - injecting a `pre_execute_hook` using `pgconnection`. This hook will prepend SQL to every query executed in the current thread. This prepended SQL sets `pgtrigger.ignore` using `SET LOCAL`.

# The issue

When the context block of `ignore` is exited, the thread-local variable is left in an empty state. However: the `pre_execute_hook` is also removed.

Furthermore, changes made by `SET LOCAL` persist for the rest of the current transaction [1].

This means:
 - The last query that is executed in the context block has a `SET LOCAL pgtrigger.ignore='{a_trigger_name}'` prepended
 - The first query that is executed after the context block has nothing prepended anymore, since the `pre_execute_hook` is now gone
 - So, assuming that we are still in the same transaction, `pgtrigger.ignore` will still be `{a_trigger_name}`, and this trigger will not run at all until either the transaction ends or a new `pgtrigger.core.ignore` context is entered and a query executed within it.

# Proposed solution

Simply emit a `SET LOCAL pgtrigger.ignore='{}';` when exiting the ignore context. This is true by definition at this point.

# Alternative hotfix

In case this approach causes problems (for example with aborted transactions that mean that we can't make queries):

Don't remove the `pre_execute_hook`, simply register it forever. It will stay around and prepend mostly useless SQL until the thread exits, but that is way better than the status quo.

[1] https://www.postgresql.org/docs/14/sql-set.html

**Comment by wesleykendall** [comment]
@ralokt thanks so much for bringing this issue up. This issue was addressed in version 2.5 (see commit [here](https://github.com/Opus10/django-pgtrigger/pull/64/commits/1501d7e21cf98ada6ffb441b6bb509fae0302e1e)). It uses an approach similar to the one suggested, calling `RESET pgtrigger.ignore` if one exits the context manager and is still inside a transaction

**Comment by ralokt** [comment]
You're most welcome, thanks for addressing it! :)
